MINIFICPP-550 - Implement RocksDB controller service and component st…

bakaid opened the pull request:
…ate storage

Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

bakaid commented:
Known parts that are missing or at least should be discussed are:
 - strategy for deleting the state of the processors that are no longer used (how do we define this? on load, reload, restart?)
 - enforcing the type of the controller service supplied for state storage
  - if we want to use KeyValueStoreServices as Processor properties we have to add a (accepts-descant-of or accepts-multiple-types kind of validator, both to minifi and to EFM
  - as the controller service used for state storage is not a Processor Property, the same kind of validation would probably not work there

phrocker commented:
@bakaid I think this is a pretty important piece of functionality especially as it relates to dealing with state more gracefully on windows (CWEL needs for example ) -- is this ready for review?

bakaid commented:
@phrocker I think it's important as well - I raised the questions in my previous comment here, because I would like input on them, but I think someone would have to see the PR before they could be answered. So yes, by all means, I would appreciate if someone could review this.

arpadboda commented on extensions/standard-processors/controllers/UnorderedMapKeyValueStoreService.h:
```diff
@@ -0,0 +1,75 @@
+
+  virtual bool get(const std::string& key, std::string& value) override;
+
+  virtual bool get(std::unordered_map<std::string, std::string>& kvs) override;
```
Let's have a pair of this to be able to store multiple in one shot!

szaszm commented on libminifi/test/TestBase.h:
```diff
@@ -304,7 +323,7 @@ class TestController {
             file += "/";
             file += dir_entry->d_name;
-            unlink(file.c_str());
+//            unlink(file.c_str());
```
This doesn't look intentional/final.

arpadboda replied:
Please note the "draft" state of the PR :)

szaszm commented on libminifi/test/integration/IntegrationBase.h:
```diff
@@ -109,6 +110,16 @@ void IntegrationBase::run(std::string test_file_location) {
+  if (mkdtemp(state_dir_buf.data()) == nullptr) {
+    throw std::runtime_error("Failed to create temporary directory for state");
+  }
+  state_dir = state_dir_buf.data();
```
duplicate code, consider extracting to a function

bakaid replied:
No longer exists in the rebased version.

szaszm commented on libminifi/test/keyvalue-tests/PersistableKeyValueStoreServiceTest.cpp:
```diff
@@ -0,0 +1,257 @@
+  const char* key = "foobar";
+  const char* value = "234";
+  const char* new_value = "baz";
+  REQUIRE(true == controller->set(key, value));
```
Comparing a boolean expression to `true` is redundant.

bakaid replied:
Yep, but in tests I think it is better readable.

arpadboda commented on libminifi/include/controllers/keyvalue/PersistableKeyValueStoreService.h:
```diff
@@ -0,0 +1,55 @@
+
+  virtual ~PersistableKeyValueStoreService();
+
+  virtual bool persist() = 0;
```
It should depend on configuration (most probably a value read from minifi.properties) in long term, so it's fine as a runtime decision.

arpadboda commented on extensions/standard-processors/controllers/UnorderedMapKeyValueStoreService.h:
```diff
@@ -0,0 +1,75 @@
+
+ protected:
+  std::unordered_map<std::string, std::string> map_;
+  std::recursive_mutex mutex_;
```
As we are mostly IO sensitive, we can afford the cost of recursive mutex.
Your point is fair, but I wish it was the narrowest bottleneck in our code... :)

bakaid commented on extensions/sftp/processors/ListSFTP.cpp:
```diff
@@ -507,85 +463,87 @@ ListSFTP::ListedEntity::ListedEntity(uint64_t timestamp_, uint64_t size_)
-    logger_->log_error("Failed to open Tracking Timestamps state file \"%s\"", tracking_timestamps_state_filename_.c_str());
-    return false;
-  }
+bool ListSFTP::updateFromTrackingTimestampsCache(const std::shared_ptr<core::ProcessContext>& context, const std::string& hostname, const std::string& username, const std::string& remote_path) {
```
Agreed, we use it because `onTrigger` already gets it this way and it is easier to just pass that to these functions.

bakaid replied:
I am not really open to in the context of this PR. This is preexisting code, and is used this way in many processors. If you want to do a round of these modifications to make easier further API changes, that makes sense, but should be a separate issue (and ticket).

bakaid commented on libminifi/include/core/ProcessContext.h:
```diff
@@ -193,6 +216,61 @@ class ProcessContext : public controller::ControllerServiceLookup, public core::
     return controller_service_provider_->getControllerServiceName(identifier);
   }
 
+  static constexpr char const* DefaultStateManagerProviderName = "defaultstatemanagerprovider";
```
> default- ok
> state- stateless is generally better, but ok
This is a PR about creating a centralized state storage mechanism for processors that require state, instead of all of them implementing individual state files. "Stateless is generally better" is a mind-bogglingly meaningless comment here.
> manager- meaningless. Only suggests doing something, so it should probably be a verb. Having a noun in place of a verb usually means trying to use a class where a function would be better.
StateManager comes from NiFi, and it is a perfectly meaningful name: it is an object that helps manage the state of a single CoreComponent. It cannot be a function as it contains some caching and validation logic that requires it in itself to be stateful.
> provider- again, a class instead of a function. This has a meaning though.
CoreComponentStateManagerProvider needs to exist as interface, which is used by the ClassLoader and we access different implementations of it through that interface.
> name- ok
> 
> I'm not saying that your code is bad, just that this is a red flag that suggests a design mistake that may or may not be in your changes.
I'm not saying your comment is useless, but it could be more useful if you tried to interpret the name in the context of the PR, instead of in itself.

phrocker replied:
Context is always important when reviewing code. Most code has a history even if we don't like it, so it may be useful to appreciate where it comes from and how we intend to get it where we want it to go.

szaszm replied:
Sorry for coming through as offensive. I lack a lot of contextual knowledge and at the time of writing the original comment, I knew even less.

I maintain my standpoint that there is most likely too much abstraction but now I understand that changing this would require redesign of significant parts of the project and it is harmful to do so without serious thought.

bakaid commented on libminifi/src/controllers/keyvalue/AbstractCoreComponentStateManagerProvider.cpp:
```diff
@@ -0,0 +1,129 @@
+    , state_valid_(false) {
+  std::string serialized;
+  if (provider_->getImpl(id_, serialized)) {
+    if (provider_->deserialize(serialized, state_)) {
```
Could you elaborate? What should be `&&`? `serialized` is not modified, `state_` is the output.

nghiaxlee replied:
I think he suggested `&&` instead of nested `if` statements

bakaid replied:
Ah, yeah, that would make sense, thank you. :D

bakaid commented on libminifi/include/core/CoreComponentState.h:
```diff
@@ -0,0 +1,62 @@
+
+  virtual bool set(const std::unordered_map<std::string, std::string>& kvs) = 0;
+
+  virtual bool get(std::unordered_map<std::string, std::string>& kvs) = 0;
```
Agreed, but signalling errors here in bools is more practical than through exceptions.

bakaid commented on libminifi/include/core/ProcessContext.h:
```diff
@@ -77,6 +85,21 @@ class ProcessContext : public controller::ControllerServiceLookup, public core::
+    if (configuration->get(minifi::Configure::nifi_state_management_provider_local, id)) {
+      auto node = controller_service_provider_->getControllerServiceNode(id);
+      if (node == nullptr) {
+        logger_->log_error("Failed to find the CoreComponentStateManagerProvider %s defined by %s", id, minifi::Configure::nifi_state_management_provider_local);
```
This is not a fatal failure, only in the case of processors that really need to store state. If state storage is misconfigured, processors that do not require it should still run fine.

bakaid commented on libminifi/test/TestBase.cpp:
```diff
@@ -30,6 +30,25 @@ TestPlan::TestPlan(std::shared_ptr<core::ContentRepository> content_repo, std::s
+  if (state_dir == nullptr) {
+    std::string state_dir_name_template = "/tmp/teststate.XXXXXX";
+    std::vector<char> state_dir_buf(state_dir_name_template.c_str(),
+                                   state_dir_name_template.c_str() + state_dir_name_template.size() + 1);
```
No longer exists in the rebased version.

bakaid commented on libminifi/src/controllers/keyvalue/AbstractAutoPersistingKeyValueStoreService.cpp:
```diff
@@ -0,0 +1,133 @@
+
+    if (!running_) {
+      logger_->log_trace("Stopping persistence thread");
+      lock.unlock();
```
True, will remove.

bakaid commented on libminifi/include/core/ProcessContext.h:
```diff
@@ -193,6 +216,61 @@ class ProcessContext : public controller::ControllerServiceLookup, public core::
+        node->enable();
+        return std::dynamic_pointer_cast<core::CoreComponentStateManagerProvider>(provider);
+      }
+    }
```
Because we need to set a File in one and a Directory in the other, obviously with different names, creating a function from this would not save us much - it would actually make it less readable in my opinion.

bakaid replied:
@szaszm Ended up adding more code to this, and now it really made sense to deduplicate it, so I've done that.

bakaid commented on libminifi/include/controllers/keyvalue/PersistableKeyValueStoreService.h:
```diff
@@ -0,0 +1,55 @@
+namespace minifi {
+namespace controllers {
+
+class PersistableKeyValueStoreService : virtual public KeyValueStoreService, public AbstractCoreComponentStateManagerProvider {
```
PersistableKeyValueStoreService is just an interface extension of PersistableKeyValueStoreService, and forcefully wrapping a PersistableKeyValueStoreService into some class that provides the CoreComponentStateManagerProvider interface after loading the  PersistableKeyValueStoreService with the classloader would make it impossible to implement a CoreComponentStateManagerProvider directly.
This way everything "just works": if you implement a PersistableKeyValueStoreService you automatically have a CoreComponentStateManagerProvider, but you can implement a CoreComponentStateManagerProvider directly.

bakaid commented:
@aboda @szaszm The PR is ready for review, with the following issues.

Blocking the PR:
 - When loading a new configuration through C2, notifyStop is not called on the ControllerServices, resulting in the RocksDB database staying open when the new controller service in the new configuration tries to reopen it, making the state retrieval fail and the processors think that they have no state. I am not sure how yet, but this could most likely be hotfixed without the whole memory leak/reload validation refactor. Will take a look next week.

Follow-up issues:
 - validation of descendant-of-* Properties: since in the new version of this PR the default injected `CoreComponentStateManagerProvider` is configurable through `minifi.properties` (whether it should be always persisting, and if not, what should be the auto persistence interval), the possibility of using your own controller service for this (defined in `config.yml`) becomes not that important for the time being. You can still do it, it will just lack a preemptive validation. Follow-up issue created: https://issues.apache.org/jira/browse/MINIFICPP-1173

 - cleaning up state storage: this is a hard question. I am not sure when we want to clean up state storage at all. For example, if a new configuration is loaded, we might want to clean the state of components that no longer exist in the flow, but this would mean that a single misconfiguration would make us loose our state (and that we can't properly roll back to an older configuration, because we have lost the state of processors no longer referenced). For the time being I think it is perfectly fine not to do any state cleanup: we don't have many processors using state, we don't have many instances of those processors that do use state, and states are small strings. We can handle the at maximum few hundred states stored in our DB (and this is the most extreme example I can imagine). If it really becomes an issue for someone, they can just delete the state directory/file. Follow-up issue created: https://issues.apache.org/jira/browse/MINIFICPP-1174

 - ControllerService notifyStop and destruction on shutdown: ControllerServices don't get destructed on shutdown (known shared_ptr cycle issue), but they also don't get a notifyStop, which Processors at least get. This is an issue, because this way the state won't be persisted on shutdown. I have worked this around by including an explicit `persist` in every Processors's notifyStop that uses state, but it should be fixed properly on the long run. Follow-up issue created: https://issues.apache.org/jira/browse/MINIFICPP-1175

 - ListSFTP migration: ListSFTP was released in 0.7.0, but I don't know about it being used in the wild. I have rewrote it for the new state handling, but didn't write any migration logic. If it is acceptable, I would prefer not to, and just make it drop the old state in the new release. I have created a blocking (for the next release) follow-up issue for this: https://issues.apache.org/jira/browse/MINIFICPP-1176

 - TailFile: TailFile is messed up. I have written state migration for it, both for the legacy and new single mode and multiple mode, but TailFile itself has issues, especially with multiple mode and rollover, it should really be rewritten from the grounds up sometime after we merged this PR. Follow-up issue created: https://issues.apache.org/jira/browse/MINIFICPP-1177

All the processors that (to my knowledge) used state has been rewritten to use the new mechanism:
 - TailFile: tested manually, created automated migration tests
 - ListSFTP: tested with automated tests
 - QueryDatabaseTable: tested state migration and normal usage manually
 - ConsumeWindowsEventLog: tested state migration and normal usage manually

bakaid commented on libminifi/src/c2/C2Agent.cpp:
```diff
@@ -596,6 +619,29 @@ void C2Agent::handle_describe(const C2ContentResponse &resp) {
       }
       enqueue_c2_response(std::move(response));
     }
+  } else if (resp.name == "corecomponentstate") {
```
This has only been tested manually. Once https://github.com/apache/nifi-minifi-cpp/pull/743 is done and merged, I can probably write reasonable tests for it.

bakaid replied:
Added `C2DescribeCoreComponentStateTest` for this.

bakaid commented on libminifi/src/c2/C2Agent.cpp:
```diff
@@ -435,6 +437,27 @@ void C2Agent::handle_c2_server_response(const C2ContentResponse &resp) {
         update_sink_->drainRepositories();
         C2Payload response(Operation::ACKNOWLEDGE, resp.ident, false, true);
         enqueue_c2_response(std::move(response));
+      } else if (resp.name == "corecomponentstate") {
```
This has only been written to make sure that it fits the architecture, it hasn't even been tested manually. Once https://github.com/apache/nifi-minifi-cpp/pull/743 is done and merged, I can probably write reasonable tests for it.

bakaid replied:
Let's revisit this after rebasing.

bakaid replied:
This is unfortunately still very much non-trivial to do (unlike testing DESCRIBE, which I do now), so I've added a TODO for it as suggested.

arpadboda commented on extensions/rocksdb-repos/controllers/RocksDbPersistableKeyValueStoreService.h:
```diff
@@ -0,0 +1,86 @@
+
+  rocksdb::DB* db_;
+  rocksdb::WriteOptions default_write_options;
+  bool db_valid_;
```
Why do we need this? The pointer being null could indicate the db not being valid. 

The documentation of rocksdb::open seems to fit in this aspect: 
```// Stores nullptr in *dbptr and returns a non-OK status on error.```
(source: https://github.com/facebook/rocksdb/blob/master/include/rocksdb/db.h )

bakaid replied:
It was a remnant from the UnorderedMap version (which was implemented before this). Replaced along with using a unique_ptr.

arpadboda commented on extensions/rocksdb-repos/controllers/RocksDbPersistableKeyValueStoreService.h:
```diff
@@ -0,0 +1,86 @@
+ protected:
+  std::string directory_;
+
+  rocksdb::DB* db_;
```
Wonder if it could be a unique ptr

bakaid replied:
None of our rocksdb stuff uses a unique_ptr, but it can, rewritten.

arpadboda commented on extensions/rocksdb-repos/controllers/RocksDbPersistableKeyValueStoreService.cpp:
```diff
@@ -0,0 +1,200 @@
+    delete db_;
+  }
+  rocksdb::Options options;
+  options.create_if_missing = true;
```
We will need the buffer options to be set here the same way as we did for FF repo, otherwise the persistent storage will slowly generate the same 64 megabyte peaks we just got rid of.

bakaid replied:
Good point, this was written last summer, we didn't know this back then.

arpadboda commented on extensions/rocksdb-repos/controllers/RocksDbPersistableKeyValueStoreService.cpp:
```diff
@@ -0,0 +1,200 @@
+}
+
+bool RocksDbPersistableKeyValueStoreService::set(const std::string& key, const std::string& value) {
+  if (!db_valid_) {
```
Wonder if we could utilize the nullcheck macro here (used id nanofi)

bakaid replied:
Replaced with unique_ptr, so no longer relevant.

arpadboda commented on extensions/sftp/processors/ListSFTP.cpp:
```diff
@@ -760,140 +713,83 @@ void ListSFTP::listByTrackingTimestamps(
-    logger_->log_error("Could not found json state file path in Tracking Entities state file \"%s\"", tracking_entities_state_filename_.c_str());
+  std::unordered_map<std::string, std::string> state_map;
+  if (!state_manager_->get(state_map)) {
+    logger_->log_debug("Failed to get state from StateManager");
```
Just debug?

Sometimes it's ok, but I wonder if it should be at least info.

bakaid replied:
If the processor has no stored state yet, `state_manager_->get` will return with false.
It could be modified to return an empty map, but I think there is value in having a distinction between "we have already stored a state, we just have nothing to store" and "we have no stored state yet".

arpadboda replied:
That's completely reasonable, I put my focus on the error case: for eg. persistency is corrupted/deleted on the disk, so processors unexpectedly start from a clean state. Which worth at least an info level log message in my opinion. 
Yes, I also expect the state manager to log something in such scenario, although loading persistency shouldn't happen too frequently (onschedule phase), so I don't mind being quite verbose here.

arpadboda commented on extensions/windows-event-log/Bookmark.cpp:
```diff
@@ -222,7 +184,7 @@ bool Bookmark::getBookmarkXmlFromFile(std::wstring& bookmarkXml) {
     logger_->log_error("No '!' in bookmarXml '%ls'", bookmarkXml.c_str());
     bookmarkXml.clear();
-    return createEmptyBookmarkXmlFile();
+	return false;
```
This seems to be misindented.

bakaid replied:
Thanks, since these are Windows-only, I edited them in VS when migrating them, and well... this is the result.

szaszm replied:
If you find a simple way to disable automatic reformatting in VS, I'm interested as well. I sometimes have to recreate my windows changes on my linux editor to avoid huge diffs due to automatic reformatting.
There are a handful of options, but no obvious (to me) way to disable messing with unchanged code.

arpadboda commented on extensions/windows-event-log/Bookmark.cpp:
```diff
@@ -10,17 +11,29 @@ namespace apache {
+				utils::file::FileUtils::concat_path(bookmarkRootDir, "uuid"), uuid), "Bookmark.txt");
+
+    std::wstring bookmarkXml;
+		if (getBookmarkXmlFromFile(bookmarkXml)) {
```
Some spaces/tabs issue here and above in this functions as well

arpadboda commented on extensions/standard-processors/controllers/UnorderedMapKeyValueStoreService.h:
```diff
@@ -0,0 +1,75 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef __UNORDERED_MAP_KEY_VALUE_STORE_SERVICE_H__
```
I think this should be called "volatileKeyValueStoreService" to maintain consistency with our repos.

bakaid replied:
I can very well imagine having multiple volatile implementations of the KeyValueStoreService in the future, with different properties than this, and I didn't want to designate this *the* VolatileKeyValueStoreService.

arpadboda replied:
Okay, I'm convinced.

arpadboda commented on extensions/standard-processors/controllers/UnorderedMapPersistableKeyValueStoreService.h:
```diff
@@ -0,0 +1,87 @@
+namespace minifi {
+namespace controllers {
+
+class UnorderedMapPersistableKeyValueStoreService : public AbstractAutoPersistingKeyValueStoreService,
```
What's the motivation behind these class?

The implementation looks good, but I'm not sure if there is any value in having a maintaining a 3rd store service. 
I wonder if we could utilise a very tiny and simple 3rd party kv storage for this purpose instead of having an own implementation. 

The non-serialized makes perfect sense for "no footprint" usage, it can be configured together with volatile repos.
The rocksdb one allignes perfectly with the rocksdb repos. 

This one stands somewhere in the middle, to ensure persistency in case of not having rocksdb, although I feel that we introduce a something we will have to support for long and it doesn't really add value.

bakaid replied:
The motivation is twofold: to provide an implementation without any third party requirements, and to provide a serialized format which is human-readable and easily editable.

As for the first motivation: we don't want to make it a hard requirement to have rocksdb to use stateful processors, because that would hinder low-footprint usages, nor do I want to make any other third party a dependency for this: it would, again, introduce an unwanted requirement for low-footprint usages and we would have to find one that supports all target platforms, and maintain it, which is not a negligible maintenance burden.

For the second one: so far users could easily clear a state by deleting the state file or directory. With the rocksdb state storage this is no longer a trivial task. The C2 methods introduced here attempt to solve this problem, but until they are finalized (or if someone does not want to use, or can't use C2), they aren't really useful.
A solution for this is using this state storage method: its serialized format is trivial and editable in text format without any third party tools, making it easy to clear (or even modify) the state of processors.

Finally for the support argument: this is a trivial, and, if you take a look at `PersistableKeyValueStoreServiceTest.cpp`, well-tested implementation, that I don't think would introduce a significant maintenance burden going forward.

bakaid commented on libminifi/src/core/ConfigurableComponent.cpp:
```diff
@@ -195,6 +195,19 @@ bool ConfigurableComponent::setSupportedProperties(std::set<Property> properties
+  std::lock_guard<std::mutex> lock(configuration_mutex_);
+
+  for (auto item : properties) {
+    properties_[item.getName()] = item;
```
This function is a modified version of `ConfigurableComponent::setSupportedProperties`. Yes, the copies can be avoided both here and there, but it is completely inconsequential, as this happens only once per flow initialization.

bakaid commented on libminifi/test/keyvalue-tests/CMakeLists.txt:
```diff
@@ -0,0 +1,42 @@
+add_test(NAME UnorderedMapPersistableKeyValueStoreServiceTest COMMAND PersistableKeyValueStoreServiceTest --config-yaml "${TEST_RESOURCES}/UnorderedMapPersistableKeyValueStoreServiceTest.yml")
+if (NOT DISABLE_ROCKSDB)
+    add_test(NAME RocksdDbPersistableKeyValueStoreServiceTest COMMAND PersistableKeyValueStoreServiceTest --config-yaml "${TEST_RESOURCES}/RocksDbPersistableKeyValueStoreServiceTest.yml")
+endif()
```
Done.

bakaid commented on libminifi/test/TestBase.cpp:
```diff
@@ -45,6 +45,23 @@ TestPlan::TestPlan(std::shared_ptr<core::ContentRepository> content_repo, std::s
+  controller_services_ = std::make_shared<core::controller::ControllerServiceMap>();
+  controller_services_provider_ = std::make_shared<core::controller::StandardControllerServiceProvider>(controller_services_, nullptr, configuration_);
+  /* Inject the default state provider ahead of ProcessContext to make sure we have a unique state directory */
+  if (state_dir == nullptr) {
+    char state_dir_name_template[] = "/tmp/teststate.XXXXXX";
+    state_dir_ = utils::file::FileUtils::create_temp_directory(state_dir_name_template);
+  } else {
+    state_dir_ = state_dir;
+  }
+  state_manager_provider_ = core::ProcessContext::getOrCreateDefaultStateManagerProvider(controller_services_provider_, configuration_, state_dir_.c_str());
```
Done.

bakaid commented on libminifi/test/integration/IntegrationBase.h:
```diff
@@ -121,6 +122,11 @@ void IntegrationBase::run(std::string test_file_location) {
 
   core::YamlConfiguration yaml_config(test_repo, test_repo, content_repo, stream_factory, configuration, test_file_location);
 
+  auto controller_service_provider = yaml_ptr->getControllerServiceProvider();
+  char state_dir_name_template[] = "/tmp/integrationstate.XXXXXX";
+  state_dir = utils::file::FileUtils::create_temp_directory(state_dir_name_template);
+  core::ProcessContext::getOrCreateDefaultStateManagerProvider(controller_service_provider, configuration, state_dir.c_str());
```
Done.

bakaid commented on extensions/standard-processors/processors/TailFile.cpp:
```diff
@@ -232,20 +236,61 @@ void TailFile::parseStateFileLine(char *buf) {
-  for (file.getline(buf, BUFFER_SIZE); file.good(); file.getline(buf, BUFFER_SIZE)) {
-    parseStateFileLine(buf);
+
+  /* We could not get the state form the StateManager, try to migrate the old state file if it exists */
```
Fixed.

bakaid commented on libminifi/src/controllers/keyvalue/AbstractCoreComponentStateManagerProvider.cpp:
```diff
@@ -0,0 +1,144 @@
+  return provider_->persistImpl();
+}
+
+AbstractCoreComponentStateManagerProvider::~AbstractCoreComponentStateManagerProvider() {
+}
```
Done.

bakaid commented on extensions/sql/processors/ExecuteSQL.cpp:
```diff
@@ -81,14 +81,14 @@ void ExecuteSQL::initialize() {
 }
 
-void ExecuteSQL::processOnTrigger(core::ProcessSession &session) {
+void ExecuteSQL::processOnTrigger(core::ProcessContext& /*context*/, core::ProcessSession& session) {
```
In a previous iteration QueryDatabaseTables's processOnTrigger needed the context, so it had to be put on the CRTP interface of SQLProcessor. It has been rewritten since and it is no longer needed by QDT, so I removed it.

bakaid commented on extensions/standard-processors/tests/unit/TailFileTests.cpp:
```diff
@@ -501,7 +481,7 @@ TEST_CASE("TailFileWithDelimiterMultipleDelimiters", "[tailfiletest2]") {
   std::mt19937 gen(std::random_device { }());
   std::generate_n(line1.begin(), 4095, [&]() -> char {
-    return 32 + gen() % (127 - 32);
+  return 32 + gen() % (127 - 32);
```
I am not really.

bakaid commented on extensions/standard-processors/controllers/UnorderedMapKeyValueStoreService.cpp:
```diff
@@ -0,0 +1,107 @@
+    exists = true;
+    value = it->second;
+  }
+  try {
+    if (!update_func(exists, value)) {
+      return false;
+    }
+  } catch (...) {
+    return false;
+  }
```
We are not silently swallowing it, we are making the operation fail, but added logging to the exception path.

bakaid commented on extensions/sftp/processors/ListSFTP.cpp:
```diff
@@ -507,85 +468,74 @@ ListSFTP::ListedEntity::ListedEntity(uint64_t timestamp_, uint64_t size_)
-        return false;
-      }
-    } else if (key == "processed.timestamp") {
-      try {
-        state_processed_timestamp = stoull(value);
-      } catch (...) {
-        logger_->log_error("processed.timestamp is not an uint64 in Tracking Timestamps state file \"%s\"", tracking_timestamps_state_filename_.c_str());
-        return false;
+    try {
+      state_processed_timestamp = stoull(state_map.at("processed.timestamp"));
+    } catch (...) {
+      return false;
+    }
```
Added individual logging.

bakaid commented on extensions/sftp/processors/ListSFTP.cpp:
```diff
@@ -760,140 +708,83 @@ void ListSFTP::listByTrackingTimestamps(
+                                            std::forward_as_tuple(name),
+                                            std::forward_as_tuple(timestamp, size));
+      } catch (...) {
         continue;
```
Added logging.

bakaid commented on extensions/sql/processors/QueryDatabaseTable.cpp:
```diff
@@ -308,19 +240,43 @@ void QueryDatabaseTable::processOnSchedule(const core::ProcessContext &context)
+      }
+    }
+  } else {
+    // Try to migrate legacy state file
+    std::string stateDir;
+    context.getProperty(s_stateDirectory.getName(), stateDir);
+    if (!stateDir.empty()) {
+      LegacyState legacyState(tableName_, stateDir, getUUIDStr(), logger_);
+      if (legacyState) {
+        mapState_ = legacyState.getStateMap();
+        if (saveState() && state_manager_->persist()) {
+          logger_->log_info("State migration successful");
+          legacyState.moveStateFileToMigrated();
+        } else {
+          logger_->log_warn("Failed to persists migrated state");
+        }
+      } else {
+        logger_->log_warn("Could not migrate state from specified State Directory %s", stateDir);
+      }
+    }
```
It is short and very much unlikely to be reused, so I don't see much benefit in moving to a function.

bakaid commented:
@arpadboda @szaszm Rebased to latest master (including C2 refactor), added test for C2 DESCRIBE corecomponentstate, addressed review comments.
I have tested state handling on shutdown and C2 configuration reload with TailFile, it seems to work fine.
I am going to continue with final verification of functionality/state migration of stateful processors on different platforms, if those work fine I think this should be ready to merge.

bakaid commented:
I am done with my final verification, the following has been tested successfully:
 - Using the new state storage for:
  - TailFile
  - ListSFTP
  - CWEL
  - QDT
 - State migration from old state file/directory for:
  - TailFile
  - CWEL
  - QDT
 - State-related minifi properties:
  - setting nifi.state.management.provider.local.always.persist to true persists every state change to disk immediately
  - setting nifi.state.management.provider.local.auto.persistence.interval to 5 sec makes state changes persist to the disk more quickly
  - using a custom configured RocksDbPersistableKeyValueStoreService through nifi.state.management.provider.local works properly
 - With both a rocksdb-enabled (default injected state storage RocksDbPersistableKeyValueStoreService) and rocksdb-disabled (default injected state storage UnorderedMapPersistableKeyValueStoreService) the following:
  - default injected storage works properly
  - state is persisted on normal shutdown
  - state is persisted on C2 configuration reload

arpadboda commented:
Built and tested, seems to work based on my tests as well, so LGTM, merging this.
Follow-ups may naturally pop up.
